src/store/addressDiscoverySlice.ts

```diff
@@ -0,0 +1,207 @@
+import { sleep } from '../utils/misc'
+import { addressesAdded } from './addressesSlice'
+import { RootState } from './store'
+
+const sliceName = 'addressDiscovery'
+
+export type DiscoveredAddress = AddressKeyPair & { balance: AddressInfo['balance'] }
+
+interface AddressDiscoveryState extends EntityState<DiscoveredAddress> {
+  loading: boolean
+  status: 'idle' | 'started' | 'stopped' | 'finished'
+}
+
+const addressDiscoveryAdapter = createEntityAdapter<DiscoveredAddress>({
+  selectId: (address) => address.hash
+})
+
+const initialState: AddressDiscoveryState = addressDiscoveryAdapter.getInitialState({
+  loading: false,
+  status: 'idle'
+})
+
+type AddressDiscoveryGroupData = {
+  highestIndex: AddressIndex | undefined
+  gap: number
+}
+
+const initializeAddressDiscoveryGroupsData = (addresses: Address[]): AddressDiscoveryGroupData[] => {
+  const groupsData: AddressDiscoveryGroupData[] = Array.from({ length: TOTAL_NUMBER_OF_GROUPS }, () => ({
+    highestIndex: undefined,
+    gap: 0
+  }))
+
+  for (const address of addresses) {
+    const groupData = groupsData[address.group]
+
+    if (groupData.highestIndex === undefined || groupData.highestIndex < address.index) {
+      groupData.highestIndex = address.index
+    }
+  }
+
+  return groupsData
+}
+
+export const addressesDiscovered = createAsyncThunk(
+  `${sliceName}/addressesDiscovered`,
+  async (_, { getState, dispatch }) => {
+    dispatch(addressDiscoveryStarted())
+
+    const minGap = 5
+    const state = getState() as RootState
+    await sleep(1) // Allow execution to continue to not block rendering
+    const { masterKey } = await walletImportAsyncUnsafe(mnemonicToSeed, state.activeWallet.mnemonic)
+    const addresses = Object.values(state.addresses.entities) as Address[]
+    const activeAddressIndexes: AddressIndex[] = addresses.map((address) => address.index)
+    const groupsData = initializeAddressDiscoveryGroupsData(addresses)
+    const derivedDataCache = new Map<AddressIndex, AddressKeyPair & { group: number }>()
+
+    let group = 0
+    let checkedIndexes = Array.from(activeAddressIndexes)
+    let maxIndexBeforeFirstGap = findMaxIndexBeforeFirstGap(activeAddressIndexes)
+
+    try {
+      while (group < 4) {
+        const groupData = groupsData[group]
+        let newAddressGroup: number | undefined = undefined
+        let index = groupData.highestIndex ?? maxIndexBeforeFirstGap
+        let newAddressData: AddressKeyPair | undefined = undefined
+
+        while (newAddressGroup !== group) {
+          index = findNextAvailableAddressIndex(index, checkedIndexes)
+          checkedIndexes.push(index)
+
+          const cachedData = derivedDataCache.get(index)
+
+          if (cachedData) {
+            if (cachedData.group !== group) {
+              continue
+            }
+
+            newAddressData = cachedData
+            newAddressGroup = cachedData.group
+          } else {
+            await sleep(1)
```

Comment:
I've probably missed that in my previous review of this algo, but why is this sleep needed?

Reply:
This is done to allow the UI to react to user input while the address discovery is still taking place. This basically stops the execution of this function to allow other JS code to run (like the handling of a button press). I use it just before a blocking expensive operation. It is not necessary for discovering addresses, but I feel like it provides a nicer UX.

Reply:
I had actually written a small comment on the other usage, but not on this one 😅 

https://github.com/alephium/mobile-wallet/blob/address-discovery/src/store/addressDiscoverySlice.ts#L84